### PR TITLE
sys/{tsrb, oorb}: Thread-safe ringbuffers

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Rename the current `tsrb` implementation to `oorb`
    - This highlights that this is a ***o***ne reader, ***o***ne writer thread-safe ***r***ing***b***uffer
- Re-add the original `tsrb` code, but insert `irq_disable()` ... `irq_restore()` where needed to provide thready-safety for multiple readers and writers
    - Existing users will just use this instead, as the API is untouched
- Add a note of warning to `oorb`
    - This issues this warns about are hopefully addressed soonish

### Testing procedure

E.g. stdio over UART on STM32 should still work. And now concurrent stdio should not result in corrupting the output (only in mixing it).

### Issues/PRs references

Addresses https://github.com/RIOT-OS/RIOT/issues/9882. (But the same issues still apply to the new `oorb`. However, with the documentation pointing those issues out, I think we can consider them as features :-D. And those "features" should be addressed in a follow up soonish.)

### CODE REVIEW

File: sys/include/oorb.h [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+#endif
+
+/**
+ * @brief     thread-safe ringbuffer struct
+ */
+typedef struct oorb {
+    uint8_t *buf;               /**< Buffer to operate on. */
+    unsigned int size;          /**< Size of buffer, must be power of 2. */
+    volatile unsigned reads;    /**< total number of reads */
+    volatile unsigned writes;   /**< total number of writes */
```
Comment (benpicco) [inline sys/include/oorb.h:52]: Do those need to be `volatie`?`

**Review by benpicco — COMMENTED** [review]
I wonder if the TSRB could just do e.g.

```C
static inline int tsrb_empty(const oorb_t *rb)
{
    int retval;
    unsigned status = irq_disable();
    retval = oorb_empty(rb);
    irq_restore(status);
    return retval;
}
```

to avoid code duplication

File: sys/include/oorb.h (hunk 1 above)
Comment (maribu) [reply sys/include/oorb.h:52]: No, those should be atomic. Or more precisely, semi-atomic :-D

(As this is the two-threads-one-buffer mode, only the reader will write to `reads` and only the writer will write to `writes`. So as long as the read and the write access to them is atomically, everything is fine. This means explicitly that a read-modify-write operation could consist of three operations with only the read and write being atomic. Real atomic accesses would make the whole read-modify-write atomically. The atomic utils helper functions introduce just the semi-atomic functions this needs - which at least for word sized variables should be lock free.)

But I intentionally left the implementation of oorb unchanged. Fixing it efficiently would require the atomic utils PR being merged. And fixing it inefficiently is done in tsrb.

**Comment by maribu** [comment]
Btw: As boths contributions have been split out, there is no need to keep this open.

**Pull request closed without merging**